File: python/chap7/detect_cycle.py
#!/usr/bin/env python

'''
Problem 7.2
Detect cycle in linked list
'''

def find_cycle(head):
    if not head or not head.next:
        return None
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            slow = head
            while slow is not fast:
                slow = slow.next
                fast = fast.next
            return slow
    return None

File: java/src/main/java/Chapter7/DetectCycle.java
package Chapter7;

import utils.ListNode;

/**
 * Created by Carl Shen on 3/24/15.
 * Problem 7.2
 * Check whether a linked list contains cycle
 */
public class DetectCycle {
    public ListNode findCycle (ListNode head) {
        if (head == null || head.next == null) return null;
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                slow = head;
                while (slow != fast) {
                    slow = slow.next;
                    fast = fast.next;
                }
                return slow;
            }
        }
        return null;
    }
}
